gems/smithy-schema/lib/smithy-schema/document.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+require_relative 'document_utils'
+
+module Smithy
+  module Schema
+    # A Smithy document type, representing typed or untyped data from Smithy data model.
+    # ## Document types
+    # Document types are protocol-agnostic view of untyped data. They could be combined
+    # with a schema to serialize its contents.
+    #
+    # Smithy-Ruby currently only support JSON documents.
+    class Document
+      # @param  [Object] data  document data
+      # @param [Hash] options
+      # @option options [Smithy::Schema::Structure] :schema schema to reference when setting
+      #  document data. Only applicable when data param is a type of {Shapes::StructureShape}.
+      # @option options [Boolean] :use_timestamp_format Whether to use the `timestampFormat`
+      #  trait or ignore it when creating a {Document} with given schema. The `timestampFormat`
+      #  trait is ignored by default.
+      # @option options [Boolean] :use_json_name Whether to use the `jsonName` trait or ignore
+      #  it when creating a {Document} with given schema. The `jsonName` trait is ignored
+      #  by default.
```

Comment:
I was pretty torn on whether I want to have these options around. These options only applies when we create a document with a given schema to decide whether we want to honor specific traits in the `@data`

My idea is that once `@data` is set - we can simply do `JSON.dump(document.data)` to load the data for a request - in lieu of having to iterate over the document data to adhere to traits during SERDE process.